src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostConfiguration.cs

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Hosting
+{
+    /// <summary>
+    /// WebAssemblyHostConfiguration is a class that implements the interface of an IConfiguration,
+    /// IConfigurationRoot, and IConfigurationBuilder. It can be used to simulatneously build
+    /// and read from a configuration object.
+    /// </summary>
+    public class WebAssemblyHostConfiguration : IConfiguration, IConfigurationRoot, IConfigurationBuilder
+    {
+        private readonly List<IConfigurationProvider> _providers = new List<IConfigurationProvider>();
+        private readonly List<IConfigurationSource> _sources = new List<IConfigurationSource>();
+
+        private readonly List<IDisposable> _changeTokenRegistrations = new List<IDisposable>();
+        private ConfigurationReloadToken _changeToken = new ConfigurationReloadToken();
+
+        /// <summary>
+        /// Returns the sources used to obtain configuration values.
+        /// </summary>
+        public IList<IConfigurationSource> Sources => _sources.ToList();
```

Comment:
I didn't follow the full discussion between you and @pranavkm (I was making chili). It looks like y'all settled on shallow copy. But if you want to spice this even more, you might think about returning a readonly wrapper here.

```C#
public IList<IConfigurationSource> Sources => new ReadOnlyCollection<IConfigurationSource>(_sources.ToList());
```

Then if someone tries to read they will get an exception instead of a silent failure. 

I'd also suggest making this property an explicit interface implementation:

```C#
IList<IConfigurationSource> IConfigurationBuilder.Sources => new ReadOnlyCollection<IConfigurationSource>(_sources.ToList());
```

That makes it *hidden* in a VIP only area. 

---

The fact that this has to exist is an unfortunate smell. This interface is weird for having both a public mutable collection and an `Add` method.

Reply:
Done and done.